gems/aws-sdk-core/lib/aws-sdk-core/instance_profile_credentials.rb

```diff
@@ -97,260 +76,128 @@ def initialize(options = {})
-          )
-          @token = Token.new(token_value, ttl, created_time) if token_value && ttl
+    def fetch_credentials
+      return '{}' if @disable_ec2_metadata
+
+      profile_name =
+        if @ec2_instance_profile_name
+          @ec2_instance_profile_name
+        else
+          metadata = @client.get(METADATA_PATH_BASE)
+          metadata.lines.first.strip
         end
-      end
-    rescue *NETWORK_ERRORS
-      # token attempt failed, reset token
-      # fallback to non-token mode
-      @token = nil
-      @imds_v1_fallback = true
-    end
 
-    # token is optional - if nil, uses v1 (insecure) flow
-    def _get_credentials(conn, token)
-      metadata = http_get(conn, METADATA_PATH_BASE, token)
-      profile_name = metadata.lines.first.strip
-      http_get(conn, METADATA_PATH_BASE + profile_name, token)
-    rescue TokenExpiredError
-      # Token has expired, reset it
-      # The next retry should fetch it
-      @token = nil
-      @imds_v1_fallback = false
-      raise Non200Response
+      @client.get(METADATA_PATH_BASE + profile_name)
+    rescue StandardError => e
+      warn("Error retrieving instance profile credentials: #{e}")
+      '{}'
     end
 
-    def token_set?
-      @token && !@token.expired?
-    end
+    def update_credentials(creds)
+      @credentials = Credentials.new(
+        creds['AccessKeyId'],
+        creds['SecretAccessKey'],
+        creds['Token']
+      )
+      @expiration =
+        creds['Expiration'] ? Time.iso8601(creds['Expiration']) : nil
+      return unless @expiration && @expiration < Time.now
 
-    def _metadata_disabled?
-      ENV.fetch('AWS_EC2_METADATA_DISABLED', 'false').downcase == 'true'
+      @no_refresh_until = Time.now + refresh_offset
+      warn_expired_credentials
     end
 
-    def open_connection
-      uri = URI.parse(@endpoint)
-      http = Net::HTTP.new(uri.hostname || @endpoint, uri.port || @port)
-      http.open_timeout = @http_open_timeout
-      http.read_timeout = @http_read_timeout
-      http.set_debug_output(@http_debug_output) if @http_debug_output
-      http.start
-      yield(http).tap { http.finish }
+    def warn_expired_credentials
+      warn('Attempting credential expiration extension due to a credential '\
+        'service availability issue. A refresh of these credentials '\
+        'will be attempted again in 5 minutes.')
     end
 
-    # GET request fetch profile and credentials
-    def http_get(connection, path, token = nil)
-      headers = { 'User-Agent' => "aws-sdk-ruby3/#{CORE_GEM_VERSION}" }
-      headers['x-aws-ec2-metadata-token'] = token if token
-      response = connection.request(Net::HTTP::Get.new(path, headers))
-
-      case response.code.to_i
-      when 200
-        response.body
-      when 401
-        raise TokenExpiredError
-      else
-        raise Non200Response
-      end
+    def empty_credentials?(creds)
+      !creds || !creds.access_key_id || creds.access_key_id.empty?
```

Comment:
You can possibly check creds.set? - it also checks secret access key but I think that's reasonable.